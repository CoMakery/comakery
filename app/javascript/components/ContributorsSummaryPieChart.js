import React from 'react'
import PropTypes from 'prop-types'
import D3Pie from 'd3pie'

class ContributorsSummaryPieChart extends React.Component {
  constructor(props) {
    super(props)
    this.createPieChart = this.createPieChart.bind(this)
    this.destroyPieChart = this.destroyPieChart.bind(this)
  }

  componentDidMount() {
    this.createPieChart()
  }

  componentWillUnmount() {
    this.destroyPieChart()
  }

  createPieChart() {
    if (this.props.chartData.length) {
      this.pieChart = new D3Pie('#award-percentages', {
        'misc': {
          'colors': {
            'background': null,
            'segments'  : [
              '#7B00D7',
              '#E5004F',
              '#0884FF',
              '#73C30E',
              '#D5E301',
              '#F6A504',
              '#C500FF',
              '#00C3EB',
              '#F85900',
              '#b00000',
              '#e4e400',
              '#baba00',
              '#878700',
              '#00b000',
              '#008700',
              '#00ffff',
              '#00b0b0',
              '#008787',
              '#b0b0ff',
              '#8484ff',
              '#4949ff',
              '#0000ff',
              '#ff00ff',
              '#b000b0'
            ],
            'segmentStroke': '#ffffff'
          }
        },
        'size': {
          'canvasHeight'  : 300,
          'canvasWidth'   : 450,
          'pieOuterRadius': '80%'
        },
        'labels': {
          'outer': {
            'pieDistance': 14
          },
          'inner': {
            'format': 'percentage'
          },
          'mainLabel': {
            'fontSize': 12,
          },
          'percentage': {
            'color'        : '#e1e1e1',
            'decimalPlaces': 2
          },
          'value': {
            'color': '#e1e1e1',
          },
          'lines': {
            'enabled': true,
            'color'  : '#cccccc'
          },
          'truncation': {
            'enabled': true
          }
        },
        'tooltips': {
          'enabled': true,
          'type'   : 'placeholder',
          'string' : '{label}: {percentage}%'
        },
        'effects': {
          'pullOutSegmentOnClick': {
            'effect': 'linear',
            'speed' : 400,
            'size'  : 10
          }
        },
        'data': {
          'content': this.props.chartData
        }
      })
    }
  }

  destroyPieChart() {
    if (this.pieChart) {
      this.pieChart.destroy()
    }
  }

  render() {
    return (
      <React.Fragment>
        <div className="columns large-4 medium-12 summary float-left">
          <h3>Total Tokens Awarded To Contributors</h3>
          <div id="award-percentages" className="royalty-pie" />
        </div>
      </React.Fragment>
    )
  }
}

ContributorsSummaryPieChart.propTypes = {
  chartData: PropTypes.array.isRequired
}
ContributorsSummaryPieChart.defaultProps = {
  chartData: []
}
export default ContributorsSummaryPieChart
