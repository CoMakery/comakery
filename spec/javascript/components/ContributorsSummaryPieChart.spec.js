import React from 'react'
import D3Pie from 'd3pie'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import ContributorsSummaryPieChart from 'components/ContributorsSummaryPieChart'

describe('Contributors Summary Pie Chart', () => {
    it('renders component correctly', () => {
      const component = renderer.create(<ContributorsSummaryPieChart/>)
      expect(component.toJSON()).toMatchSnapshot()
      // D3Pie relies on D3 which requires an actual DOM to operate,
      // so we are not able to access the chart in this test environment.
      // D3 can be fooled by faux DOM such as https://github.com/Olical/react-faux-dom,
      // but it looks kinda hacky and in our case might be an overkill.
    })
    it('has correct header', () => {
      const component = shallow(<ContributorsSummaryPieChart/>)
      expect(component.find('h3').text()).toEqual('Total Tokens Awarded To Contributors')
    })
})