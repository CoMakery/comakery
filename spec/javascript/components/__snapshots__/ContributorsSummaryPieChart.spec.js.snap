// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contributors Summary Pie Chart renders component correctly 1`] = `
<Fragment>
  <div
    className="columns large-4 medium-12 summary float-left"
  >
    <h3>
      Total Tokens Awarded To Contributors
    </h3>
    <div
      className="royalty-pie"
      id="award-percentages"
    />
  </div>
</Fragment>
`;
